src/main/java/com/minecolonies/entity/ai/citizen/guard/AbstractEntityAIGuard.java

```diff
@@ -290,17 +387,73 @@ protected int getMaxAttacksUntilRestock()
     protected AIState patrol()
     {
         worker.setAIMoveSpeed(1);
+        AbstractBuilding building = getOwnBuilding();
 
-        if (currentPatrolTarget == null)
+        if (building != null && building instanceof BuildingGuardTower)
         {
-            currentPatrolTarget = getRandomBuilding();
+            if (currentPatrolTarget == null
+                    || BlockPosUtil.getDistance2D(building.getColony().getCenter(), currentPatrolTarget) > Configurations.workingRangeTownHall + Configurations.townHallPadding
+                    || currentPatrolTarget.getY() < 5)
+            {
+                return getNextPatrollingTarget((BuildingGuardTower) building);
+            }
+
+            if (worker.isWorkerAtSiteWithMove(currentPatrolTarget, PATH_CLOSE))
+            {
+                return getNextPatrollingTarget((BuildingGuardTower) building);
+            }
         }
 
-        if (worker.isWorkerAtSiteWithMove(currentPatrolTarget, PATH_CLOSE))
+        return AIState.GUARD_SEARCH_TARGET;
+    }
+
+    /**
+     * Retrieves the next patrolling target from the guard tower.
+     * @param building his building.
+     * @return the next state to go to.
+     */
+    private AIState getNextPatrollingTarget(BuildingGuardTower building)
+    {
+        if (building.shallPatrolManually() && building.getTask().equals(BuildingGuardTower.Task.PATROL))
         {
-            currentPatrolTarget = null;
+            BlockPos pos = building.getNextPatrolTarget(currentPatrolTarget);
```

Comment:
![MINOR](https://raw.githubusercontent.com/SonarCommunity/sonar-github/master/images/severity-minor.png) Local variable 'pos' could be declared final [![rule](https://raw.githubusercontent.com/SonarCommunity/sonar-github/master/images/rule.png)](http://home.kk-sc.de:9000/coding_rules#rule_key=pmd%3ALocalVariableCouldBeFinal)